File: analyzers/tests/SonarAnalyzer.Test/TestCases/PartCreationPolicyShouldBeUsedWithExportAttribute.vb
Imports System
Imports System.ComponentModel.Composition

Namespace Tests.TestCases
    <Export(GetType(Object))>
    <PartCreationPolicy(CreationPolicy.Any)> ' Compliant, Export is present
    class Program1

    End Class

    <InheritedExport(GetType(Object))>
    <PartCreationPolicy(CreationPolicy.Any)> ' Compliant, InheritedExport is present
    class Program2_Base

    End Class

    <PartCreationPolicy(CreationPolicy.Any)> ' Compliant, InheritedExport is present in base
    class Program2
        inherits Program2_Base

    End Class

    <PartCreationPolicy(CreationPolicy.Any)> ' Noncompliant {{Add the 'ExportAttribute' or remove 'PartCreationPolicyAttribute' to/from this type definition.}}
    class Program3
'    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^@-1

    End Class

    <PartCreationPolicy(CreationPolicy.Any)> ' Noncompliant, Export is not inherited
    class Program4
        inherits Program1

    End Class

    <MyExportAttribute>
    <PartCreationPolicy(CreationPolicy.Any)>
    Class Program6
    End Class

    <MyInheritedExport(GetType(Object))>
    <PartCreationPolicy(CreationPolicy.Any)>
    Class Program7_Base
    End Class

    <PartCreationPolicy(CreationPolicy.Any)>
    Class Program7
        Inherits Program2_Base
    End Class

    <Foo>
    <PartCreationPolicy(CreationPolicy.Any)> ' Noncompliant {{Add the 'ExportAttribute' or remove 'PartCreationPolicyAttribute' to/from this type definition.}}
    Class Program8
    End Class

    <PartCreationPolicy(CreationPolicy.NonShared)>
    Class Program9
        Implements IMyInheritedExportInterface
    End Class

    <PartCreationPolicy(CreationPolicy.[Shared])>
    Class Program11
        Implements IFoo, IBar, IMyInheritedExportInterface, IQix
    End Class

    <PartCreationPolicy(CreationPolicy.[NonShared])> ' Noncompliant
    Class Program12
        Implements IFoo, IBar, IQix
    End Class

    Class MyExportAttribute
        Inherits ExportAttribute
    End Class

    Class MyInheritedExportAttribute
        Inherits InheritedExportAttribute

        Public Sub New(ByVal type As System.Type)
        End Sub
    End Class

    Class FooAttribute
        Inherits Attribute
    End Class

    <InheritedExport(GetType(Object))>
    Interface IMyInheritedExportInterface
    End Interface

    Interface IFoo
    End Interface

    Interface IBar
    End Interface

    Interface IQix
    End Interface

End Namespace

File: analyzers/tests/SonarAnalyzer.Test/TestCases/PartCreationPolicyShouldBeUsedWithExportAttribute.cs
using System;
using System.ComponentModel.Composition;

[Export(typeof(object))]
[PartCreationPolicy(CreationPolicy.Any)] // Compliant, Export is present
class Program1
{
}

[InheritedExport(typeof(object))]
[PartCreationPolicy(CreationPolicy.Any)] // Compliant, InheritedExport is present
class Program2_Base
{
}

[PartCreationPolicy(CreationPolicy.Any)] // Compliant, InheritedExport is present in base
class Program2 : Program2_Base
{
}

    [PartCreationPolicy(CreationPolicy.Any)] // Noncompliant {{Add the 'ExportAttribute' or remove 'PartCreationPolicyAttribute' to/from this type definition.}}
//   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
class Program3
{
}

[PartCreationPolicy(CreationPolicy.Any)] // Noncompliant, Export is not inherited
class Program4 : Program1
{
}

class Program5
{
    [PartCreationPolicy(CreationPolicy.Any)] // Error [CS0592] - Compliant, attribute cannot be used on methods, don't raise
    public void Method() { }

    [PartCreationPolicy(CreationPolicy.Any)] // Error [CS0592] - Compliant, attribute cannot be used on fields, don't raise
    public int Field;

    [PartCreationPolicy(CreationPolicy.Any)] // Error [CS0592] - Compliant, attribute cannot be used on properties, don't raise
    public int Property { get; set; }
}

[MyExport]
[PartCreationPolicy(CreationPolicy.Any)] // Compliant, custom Export is present
class Program6 { }

[MyInheritedExport(typeof(object))]
[PartCreationPolicy(CreationPolicy.Any)] // Compliant, MyInheritedExport is present
class Program7_Base { }

[PartCreationPolicy(CreationPolicy.Any)] // Compliant, MyInheritedExport is present in base
class Program7 : Program2_Base { }

[Foo]
[PartCreationPolicy(CreationPolicy.Any)] // Noncompliant {{Add the 'ExportAttribute' or remove 'PartCreationPolicyAttribute' to/from this type definition.}}
class Program8 { }

[PartCreationPolicy(CreationPolicy.NonShared)] // Compliant, InheritedExport is present on interface
class Program9 : IMyInheritedExportInterface { }

[PartCreationPolicy(CreationPolicy.Shared)] // Compliant, MyInheritedExport is present on interface MyInheritedExportInterface
class Program11 : IFoo, IBar, IMyInheritedExportInterface, IQix { }

[PartCreationPolicy(CreationPolicy.NonShared)] // Noncompliant
class Program12 : IFoo, IBar, IQix { }

class MyExportAttribute : ExportAttribute { }

class MyInheritedExportAttribute : InheritedExportAttribute
{
    public MyInheritedExportAttribute(System.Type type) { }
}

class FooAttribute : Attribute { }

[InheritedExport(typeof(object))]
interface IMyInheritedExportInterface { }

interface IFoo { }
interface IBar { }
interface IQix { }

[PartCreationPolicy(CreationPolicy.Any)] // Error [CS1671] - Compliant, illegal use, don't raise

namespace Aliases
{
    using AliasForPartCreationPolicy = System.ComponentModel.Composition.PartCreationPolicyAttribute;
    using AliasForExport = System.ComponentModel.Composition.ExportAttribute;
    using AliasForInheritedExport = System.ComponentModel.Composition.InheritedExportAttribute;

    [AliasForExport(typeof(object))]
    [PartCreationPolicy(CreationPolicy.Any)]         // Compliant, Export is present
    class AClass1 { }

    [AliasForInheritedExport(typeof(object))]
    [PartCreationPolicy(CreationPolicy.Any)]         // Compliant, InheritanceExport is present
    class AClass2 { }

    [AliasForExport(typeof(object))]
    [AliasForPartCreationPolicy(CreationPolicy.Any)] // Compliant, Export is present
    class AClass3 { }

    [AliasForPartCreationPolicy(CreationPolicy.Any)] // Noncompliant
    class AClass4 { }
}
